src/accelerate/accelerator.py

```diff
@@ -2120,8 +2121,17 @@ def gather_for_metrics(self, tensor):
         9
         ```
         """
-        tensor = self.gather(tensor)
 
+        try:
+            recursively_apply(lambda x: x, input_data, error_on_other_type=True)
+            all_tensors = True
+        except TypeError:
+            all_tensors = False
+
+        if not all_tensors:
+            data= gather_object(input_data)
+
+        data= self.gather(input_data)
```

Comment:
We need an else here, sorry probably did a poor suggestion :) 

```suggestion
        if not all_tensors:
            data = gather_object(input_data)
        else:
            data = self.gather(input_data)
```